src/OrleansRuntime/GrainTypeManager/GrainTypeManager.cs

```diff
@@ -151,24 +149,27 @@ internal IReadOnlyList<SiloAddress> GetSupportedSilos(int typeCode)
             return supportedSilosByTypeCode[typeCode];
         }
 
-        internal IReadOnlyList<SiloAddress> GetSupportedSilos(int typeCode, int ifaceId, ushort version)
+        internal IReadOnlyList<SiloAddress> GetSupportedSilos(int typeCode, int ifaceId, IReadOnlyList<ushort> versions)
         {
-            var key = Tuple.Create(typeCode, ifaceId, version);
-            return supportedSilosCache.GetOrAdd(
-                key,
-                tuple =>
-                {
-                    var silosWithTypeCode = supportedSilosByTypeCode[typeCode];
-                    return supportedSilosByInterface[ifaceId][version].Intersect(silosWithTypeCode).ToList();
-                });
+            var result = new List<SiloAddress>();
+            foreach (var version in versions)
```

Comment:
I'm sure I sound like a broken record, always complaining about allocations, but if `versions` used the concrete type instead of an interface, then this might not allocate an enumerator.